docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,694 @@
+
+#### `MeshGateway`
+
+This new resource is intended to replace the `MeshGatewayRoute` resource for
+configuring routes for a `MeshGateway`.
+
+There are two different ways we can express a `MeshGateway` attachment either as
+`spec.targetRef` & `to.targetRef` and `from.targetRef` & `spec.targetRef`.
+
+The core difference is that routes for a `MeshGateway` are fundamentally different from in-mesh
+routes because there's a proxy "in between" requests. That is, when I send a
+request from service to service, the Envoy configuration and routing happens on the source
+proxy side. When I send a request to a `MeshGateway`, the request reaches the
+`MeshGateway` proxy and _then_ gets routed, i.e. they're on the destination
+side.
+
+##### `MeshGateway` as `to.targetRef`
+
+This is a somewhat different way of thinking about `targetRef` because here the
+`to.targetRef` points to the Envoy whose configuration is changing, as opposed
+to `spec.targetRef`. However, a `MeshGateway` proxy is fundamentally different from a
+sidecar proxy because it does not attach to a service, so perhaps this is
+justified?
+
+###### Edge gateway
+
+In order to configure a non-cross-mesh `MeshGateway` resource,
+one option is that the user points `to.targetRef` to a `MeshGateway`.
+
+```yaml
+targetRef:
+ kind: Mesh
+to:
+ - targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+```
+
+the `spec.targetRef` of `kind: Mesh` can be read as simply meaning "any
+connections made to the `MeshGateway`". Only `kind: Mesh` is permitted with a
+`to.targetRef` of `kind: MeshGateway`.
+
+###### Cross-mesh gateway
+
+For cross-mesh `MeshGateways`:
+
+```yaml
+targetRef:
+ kind: Mesh
+to:
+ - targetRef:
+    kind: MeshGateway
+    name: mesh-gateway
+```
+
+would target _any requests_ from _any `Mesh`_ made to this `MeshGateway` whereas
+`name` must be used to target a specific `Mesh` as source:
+
+```yaml
+targetRef:
+ kind: Mesh
+ name: other-mesh
+to:
+ - targetRef:
+    kind: MeshGateway
+    name: mesh-gateway
+```
+
+##### `MeshGateway` in `spec.targetRef`
+
+If we put `MeshGateway` in `spec.targetRef`, we could instead use `from`.
+
+###### Edge gateway
+
+In order to configure a non-cross-mesh `MeshGateway` resource:
+
+```yaml
+targetRef:
+ kind: MeshGateway
+ name: edge-gateway
+from:
+ - targetRef:
+    kind: Mesh
+```
+
+the `from.targetRef` of `kind: Mesh` can be read as simply meaning "any
+connections made to the `MeshGateway`". Only `kind: Mesh` is permitted as a
+`from.targetRef` with `targetRef` of `MeshGateway`.
+
+###### Cross-mesh gateway
+
+For cross-mesh `MeshGateways`:
+
+```yaml
+targetRef:
+ kind: MeshGateway
+ name: mesh-gateway
+from:
+ - targetRef:
+    kind: Mesh
+```
+
+would target _any requests_ from _any `Mesh`_ made to this `MeshGateway` whereas
+`name` must be used to target a specific `Mesh` as source:
+
+```yaml
+targetRef:
+ kind: MeshGateway
+ name: mesh-gateway
+from:
+ - targetRef:
+    kind: Mesh
+    name: other-mesh
+```
```

Comment:
I prefer this option the top-level targetRef keeps the same semantic which seems better to me.

Reply:
resolving because there's another option